### PR TITLE
[FLINK-31192][connectors/dataGen] Fix dataGen takes too long to initi…

### PR DESCRIPTION
**Pull request by xuzhiwen1255**
## What is the purpose of the change
In the sequence mode, the dataGen field will preload all the sequence values when it is initialized, and it will take a long time during this period

## Brief change log
- Adapt the way of preloading the value to generate a sequence value every time the next method is called
- The checkpoint saves not all the sequence values, but the position where the current task sends the sequence value

## Verifying this change
`DataGeneratorSourceTest` and `DataGenTableSourceFactoryTest`

## Does this pull request potentially affect one of the following parts:
  - Dependencies (does it add or upgrade a dependency):   no 
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`:   no 
  - The serializers:   no 
  - The runtime per-record code paths (performance sensitive):   no  
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper:   no 
  - The S3 file system connector:   no 

## Documentation
  - Does this pull request introduce a new feature?  no

### CODE REVIEW

**Comment by xuzhiwen1255** [comment]
@reswqa Can you review it for me, Thank you.

**Comment by reswqa** [comment]
Thanks @xuzhiwen1255 for creating this, It seems that the CI is failure, PTAL first.

**Comment by xuzhiwen1255** [comment]
@flinkbot run azure

**Comment by xuzhiwen1255** [comment]
@reswqa Re-running the cli is no problem.

**Comment by xuzhiwen1255** [comment]
@reswqa CI is fine now, can you come over and have a look?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java [hunk 1]
```diff
@@ -42,9 +44,14 @@
 
     private final long start;
     private final long end;
+    private long totalNoOfElements;
+    /**
+     * Save the intermediate state of the data to be sent by the current subtask,when the state
+     * restore, the sequence values continue to be sent based on the intermediate state.
+     */
+    private ArrayList<InternalState> internalStates;
```
Comment (RyanSkraba) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java:52]: Does this need to be transient as well?

**Review by RyanSkraba — COMMENTED** [review]
I think this is the right approach (instead of pregenerating potentially enormous lists in memory)!   But there might be an opportunity to simplify the logic -- I made a suggestion in `nextValue()` .  What do you think?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java [hunk 2]
```diff
@@ -65,33 +72,46 @@ public void open(
...
+    }
 
-            for (long collected = 0; collected < toCollect; collected++) {
-                this.valuesToEmit.add(collected * stepSize + congruence);
+    public Long nextValue() {
+        Iterator<InternalState> iterator = internalStates.iterator();
+        if (iterator.hasNext()) {
+            InternalState state = iterator.next();
+            long nextSequence = state.collected * state.stepSize + (start + state.taskId);
+            state.collected++;
```
Comment (RyanSkraba) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java:101]: If we are updating the current state at every iteration anyway: instead of incrementing `collected`, why not just retain the value in the current state?  It's easy to calculate the *next* (just add stepSize) and you know to drop the state from the list if it goes over `this.end`.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java [hunk 3]
```diff
@@ -65,33 +72,46 @@ public void open(
...
+                            safeDivide(totalNoOfElements, state.stepSize),
+                            state.stepSize,
+                            state.taskId)) {
+                iterator.remove();
             }
+            return nextSequence;
         }
+
+        // Before calling this method, you should call hasNext to check
+        throw new IllegalStateException();
```
Comment (RyanSkraba) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java:114]:
```suggestion
        throw new IllegalStateException("SequenceGenerator.nextValue() was called with no remaining values.");
```
This code should be unreachable, so it's not a big deal -- but just in case, a helpful message is a good practice.

Comment (xuzhiwen1255) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java:114]: You're quite right. I'll change it.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java (hunk 2 above)
Comment (xuzhiwen1255) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java:101]: Let me rethink, perhaps it is more reasonable to change this way.

**Comment by xuzhiwen1255** [comment]
@flinkbot run azure

**Comment by xuzhiwen1255** [comment]
@RyanSkraba  @XComp Please check again, thanks.

**Comment by xuzhiwen1255** [comment]
@flinkbot run azure

**Comment by xuzhiwen1255** [comment]
@flinkbot run azure

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java [hunk 4]
```diff
@@ -65,33 +71,40 @@ public void open(
...
-                this.valuesToEmit.add(collected * stepSize + congruence);
+    public Long nextValue() {
+        Iterator<InternalState> iterator = internalStates.iterator();
+        if (iterator.hasNext()) {
+            InternalState state = iterator.next();
+            long nextSequence =
+                    state.currentValue == 0 ? this.start + state.taskId : state.currentValue;
+            state.currentValue = nextSequence + state.stepSize;
+            // All sequence values are cleared from the stateList after they have been sent
+            if (state.currentValue > this.end) {
```
Comment (RyanSkraba) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java:99]: If this.end is close to `Long.MAX_VALUE`, there's a possibility of overflow here with `nextSequence + state.stepSize`.

```suggestion
            if (state.currentValue > this.end || state.currentValue <= nextSequence) {
```

(I think this is correct logic, but `Math.addExact` could be used as well.)

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java [hunk 5]
```diff
@@ -65,33 +71,40 @@ public void open(
...
+        }
+    }
 
-            for (long collected = 0; collected < toCollect; collected++) {
-                this.valuesToEmit.add(collected * stepSize + congruence);
+    public Long nextValue() {
+        Iterator<InternalState> iterator = internalStates.iterator();
+        if (iterator.hasNext()) {
+            InternalState state = iterator.next();
+            long nextSequence =
```
Comment (RyanSkraba) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java:95]:
```suggestion
            long nextSequence =
```

A really minor suggestion, but I'd inverse these variable names for readability!

The one stored in the InternalState is the *next* or future value (to be emitted on a later call): maybe `futureValue`?

The one currently being store in this local variable is the `currentValue`.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java [hunk 6]
```diff
@@ -65,33 +71,40 @@ public void open(
...
+                    state.currentValue == 0 ? this.start + state.taskId : state.currentValue;
```
Comment (RyanSkraba) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java:96]: More seriously, there's a problem with using `state.currentValue == 0` as the unitialized state.  Imagine if you're dumping the range from -1000L to 1000L, when the task finds itself *meant* to be emitting zero, it will start over from it's first value instead.

I'm working through it a bit, but it might be enough just to set it in the InternalState from the start:  `new InternalState(taskIdx, stepSize, start + taskIdx)`.

**Review by RyanSkraba — CHANGES_REQUESTED** [review]
Hey, this looks pretty good -- it cleaned up the code pretty nicely and uses a lot less fiddly math!  I think I found a bug though, can you take a look?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java (hunk 5 above)
Comment (xuzhiwen1255) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java:95]: How about calling it nextValue, I think it's more reasonable.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java (hunk 6 above)
Comment (xuzhiwen1255) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java:96]: Thanks for the reminder, I didn't think it was possible to use a negative number as a starting value.

**Review by RyanSkraba — APPROVED** [review]
LGTM!  This looks like a pretty nice clean-up, especially with enormous sequences.  Thanks!

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java [hunk 7]
```diff
@@ -65,33 +71,46 @@ public void open(
...
+            final long stepSize = runtimeContext.getNumberOfParallelSubtasks();
+            InternalState state = new InternalState(taskIdx, stepSize, start + taskIdx);
+            internalStates.add(state);
+        }
+    }
 
-            for (long collected = 0; collected < toCollect; collected++) {
-                this.valuesToEmit.add(collected * stepSize + congruence);
+    public Long nextValue() {
+        Iterator<InternalState> iterator = internalStates.iterator();
```
Comment (XComp) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java:92]: When do we have the case that there is more than one state in the `internalStates`? Is that when we're reducing the parallelism and multiple states get picked up by a single operator? Is it enough to always select the first `InternalState`? It feels to me that we're processing one state first until we reach the end and would then process the next state which has a way lower number again? :thinking:

**Review by XComp — COMMENTED** [review]
Thanks for coming up with this change, @xuzhiwen1255 and for reviewing it, @RyanSkraba. I have a few comments. Please see them below (disclaimer: this part of the code is not something I am familiar with; sorry if I asked too many questions :innocent: ).

In addition to the comments, I feel like we're lacking some tests here. Is there a way to unit test the generator (maybe even in the context of a parallelism change)?

**Comment by XComp** [comment]
On another note: We should rebase the branch to include FLINK-31962 which would make CI succeed again.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java [hunk 8]
```diff
@@ -65,33 +71,46 @@ public void open(
...
+            // The first time the job is executed.
             final int taskIdx = runtimeContext.getIndexOfThisSubtask();
-            final long congruence = start + taskIdx;
-
-            long totalNoOfElements = Math.abs(end - start + 1);
-            final int baseSize = safeDivide(totalNoOfElements, stepSize);
-            final int toCollect =
-                    (totalNoOfElements % stepSize > taskIdx) ? baseSize + 1 : baseSize;
+            final long stepSize = runtimeContext.getNumberOfParallelSubtasks();
+            InternalState state = new InternalState(taskIdx, stepSize, start + taskIdx);
```
[the remark this answers is not in the document]
Comment (XComp) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java:86]: I'm also wondering whether we could add some JavaDoc (either to `SequenceGenerator` or to `InternalState`) to describe the logic of this class a bit more.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java (hunk 7 above)
Comment (xuzhiwen1255) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java:92]: Yes, multiple states will only appear when the degree of parallelism is reduced.

I guess what you're thinking is: are we going to get the latest sequence value by putting the state into the queue and then taking it off the head of the queue each time and putting it at the end? Instead of sending all the sequences of a state and then sending the sequence value of the next state?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (xuzhiwen1255) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java:86]: > Why do we pass the `taskIdx` as a dedicated parameter? It doesn't look like we're using it anywhere? 🤔

Yes it is not used and i will delete it.

[the remark this answers is not in the document]
Comment (xuzhiwen1255) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java:86]: Its function is to distinguish the starting value of each degree of parallelism according to start.

I think it could be called `startOffset`, the offset used to determine the start value of each task.

[the remark this answers is not in the document]
Comment (XComp) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java:86]: I think that's a better name for it, yes. :+1:

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java (hunk 7 above)
Comment (xuzhiwen1255) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java:92]: I'll try to put it in unit tests... ..

**Comment by xuzhiwen1255** [comment]
> In addition to the comments, I feel like we're lacking some tests here. Is there a way to unit test the generator (maybe even in the context of a parallelism change)?

I think about it, the change of parallelism already exists in the existing unit test, I think we don't need to add the test, because it already includes the complete generated test case of the sequence.

> In this DataGeneratorSourceTest#innerTestDataGenCheckpointRestore test method, the degree of parallelism is modified .

**Comment by xuzhiwen1255** [comment]
@flinkbot run azure

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java (hunk 7 above)
Comment (RyanSkraba) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java]: This seems like a good idea, and it's probably appropriate to create some JIRAs to implement in subsequent, incremental steps.

The current internal states (before this PR) for a generator from 1 to 1M would have 1 million values: `1,2,3...1000000`.  If it were checkpointed, it would **(1)** take quite a bit of space and **(2)** I'm not sure that restoring it to a lower parallelism retains any lowest-first semantics.

This PR solves **(1)**, but doesn't change much in terms of **(2)**.  The internal states would be `(1,3), (2,3), (3,3)` (assuming a parallelism of 3).  

If it was restored to a parallelism of two (after the PR), and one task got `(1,3), (2,3)` as its internal states, it would emit `1,4,7,10 ... 1000000, 2, 5, 8 ...`.  

If my understanding is correct, emitting `2` _after_ `1000000` is a possibility in both scenarios.  There's an opportunity here:

* Improve the logic so that a task always emits the lowest of the possible values from the internal states that it possesses (although a priority queue is probably unnecessary, it should end up roughly equivalent to sorting and cycling through the states).  This is really a new improvement that could be applied to the current state.
* Avoid performance skew after reducing parallelism.  Currently, going from N tasks to N-1 tasks will double the length of time it takes, since one task will have twice as many values to emit as the others.  This is a undesirable consequence of this PR, but can probably also be discussed and implemented in a subsequent step.

One thing that I really like about this PR is that it enables huge sequences that aren't possible today, in an efficient manner.  This is a really compelling advantage over the current state in every way!

Comment (xuzhiwen1255) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java]: > * Avoid performance skew after reducing parallelism.  Currently, going from N tasks to N-1 tasks will double the length of time it takes, since one task will have twice as many values to emit as the others.  This is a undesirable consequence of this PR, but can probably also be discussed and implemented in a subsequent step.

Yes, I noticed this problem, but I haven't thought of a better solution.

But I have an idea, that is, for each subtask, we let him generate more InternalState, and the general process is described as follows:

Assuming that the parallelism is 5 and sending 100 pieces of data, we let each subtaks be logically divided into multiple InternalStates
The current InternalState is as follows
```java
InternalState{
         long stepSize;
         long nextValue;
}
```
When subtask-0 is initialized: only one InternalState(5,0) will be initialized;


So my idea is to add an end to indicate the end value of the current InternalState
```java
InternalState{
         long stepSize;
         long nextValue;
         long end;
}
```
Cutting when subtask-0 is initialized produces two: InternalState(5,0,50), InternalState(5,0,100);
Of course, it doesn’t have to be two, it can be multiple, so that when the parallelism is readjusted, one task can avoid processing multiple InternalStates.
@XComp @RyanSkraba  WDYT?

At that time, I will raise a jira to track this issue, and then we will discuss this issue.

Comment (XComp) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java]: Thanks for sharing your view, @RyanSkraba. I think you're right - the implementation in master (i.e. without this PR's change) doesn't support reducing the parallelism properly, too. 
```
Avoid performance skew after reducing parallelism. Currently, going from N tasks to N-1 tasks will double the length of time it takes, since one task will have twice as many values to emit as the others. This is a undesirable consequence of this PR, but can probably also be discussed and implemented in a subsequent step.
```
I don't understand that part: Isn't the performance skew issue you're describing more general and not rely connected to the `DataGenerator`? The skew you're mentioning is happening in any case when reducing the parallelism. It's more of a question how Flink would distribute the operator state after the parallelism was reduced. :thinking: Therefore, I don't see this as a problem for this PR.

**Comment by XComp** [comment]
> I think about it, the change of parallelism already exists in the existing unit test, I think we don't need to add the test, because it already includes the complete generated test case of the sequence.

Thanks for pointing that out. I missed `DataGeneratorSourceTest`. Do you still see the value in having a `SequenceGeneratorTest` testing the basic functionality of the `SequenceGenerator`, though?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java (hunk 7 above)
Comment (RyanSkraba) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java]: @xuzhiwen1255 This sounds perfect!
> But I have an idea, that is, for each subtask, we let him generate more InternalState

That is to say, instead of having N internal states when there are N tasks initially, we would have 2N, or 10N, or 100N.  That number doesn't have to be extremely high to avoid skew here, and still benefit from the "concise" and efficient InternalState you've implemented here.

I think there might be some cool way of splitting up the ranges differently, but I'd like to think about it!  This should probably go on the future JIRA though :D 

@XComp Yes, this issue would happen to any source that created "one split per task" on initialization, where the split corresponds to the operator state -- but the previous implementation didn't do that.  It was one state per element to emit, so it _could_ be perfectly rebalanced (but overall pretty inefficiently).

I agree that it's not a problem for this PR!

**Comment by xuzhiwen1255** [comment]
> > I think about it, the change of parallelism already exists in the existing unit test, I think we don't need to add the test, because it already includes the complete generated test case of the sequence.
> 
> Thanks for pointing that out. I missed `DataGeneratorSourceTest`. Do you still see the value in having a `SequenceGeneratorTest` testing the basic functionality of the `SequenceGenerator`, though? It might help to understand the contract of the `SequenceGenerator` class.


I don't think it is necessary to use `SequenceGeneratorTest` for testing, because the existing tests have covered its basic functions. Later, we may need to test how to avoid data skew and test sequence generation integrity when adjusting operator parallelism.


@XComp WDYT?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java (hunk 7 above)
Comment (xuzhiwen1255) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java]: > That is to say, instead of having N internal states when there are N tasks initially, we would have 2N, or 10N, or 100N. That number doesn't have to be extremely high to avoid skew here, and still benefit from the "concise" and efficient InternalState you've implemented here.

Yes, but I haven't figured out how to solve this problem reasonably.

**Comment by xuzhiwen1255** [comment]
@flinkbot run azure

**Comment by xzw0223** [comment]
@XComp @RyanSkraba Any more questions?

**Comment by XComp** [comment]
Hi @xuzhiwen1255 , sorry for making you wait for so long. I was on vacation and have to focus on some other work right now. I try to get back to this PR soon'ish

**Comment by xuzhiwen1255** [comment]
> Hi @xuzhiwen1255 , sorry for making you wait for so long. I was on vacation and have to focus on some other work right now. I try to get back to this PR soon'ish

@XComp  It's all right, have a nice holiday.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java [hunk 9]
```diff
@@ -22,29 +22,44 @@
 import org.apache.flink.api.common.functions.RuntimeContext;
 import org.apache.flink.api.common.state.ListState;
 import org.apache.flink.api.common.state.ListStateDescriptor;
-import org.apache.flink.api.common.typeutils.base.LongSerializer;
+import org.apache.flink.api.common.typeinfo.TypeInformation;
 import org.apache.flink.runtime.state.FunctionInitializationContext;
 import org.apache.flink.runtime.state.FunctionSnapshotContext;
 import org.apache.flink.util.Preconditions;
 
+import org.apache.flink.shaded.guava30.com.google.common.collect.Queues;
```
Comment (XComp) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java:30]:
```suggestion
import org.apache.flink.shaded.guava31.com.google.common.collect.Queues;
```
This won't work anymore because the `flink-shaded` dependencies were upgraded (FLINK-32032). Rebasing the branch might reveal other affected code locations.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java [hunk 10]
```diff
@@ -22,29 +22,44 @@
...
  * A stateful, re-scalable {@link DataGenerator} that emits each number from a given interval
  * exactly once, possibly in parallel.
+ *
+ * <p>It maintains a state internally to record the position of the current subtask sending
+ * sequence. When the task resumes, it will continue to send the sequence value according to the
+ * position sent by the state, until all the sequences have been sent.
+ *
+ * <p><b>IMPORTANT NOTE: </b> When the degree of parallelism increases, there may be cases where
+ * subtasks are running empty. When the degree of parallelism decreases, there may be cases where
+ * one subtask handles multiple states.
```
Comment (XComp) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java:49]: Shall we also add that decreasing the parallelism might result in an out-of-orderness of the sequence?

**Review by XComp — COMMENTED** [review]
Sorry for letting you wait for so long. I was focusing on the finalizing the leader election work for 1.18 (FLINK-26522). Anyway, now that this is done, I'm happy to pick this up and finalize it.

We're already in a good state with the PR, I guess. I added some minor comments. PTAL

> I don't think it is necessary to use SequenceGeneratorTest for testing, because the existing tests have covered its basic functions. Later, we may need to test how to avoid data skew and test sequence generation integrity when adjusting operator parallelism.

I still think that it might make sense to introduce a `SequenceGeneratorTest`. This would allow us to test the edge cases easily (e.g. reaching `Long.MAX_VALUE`) more specifically. The only test, I found was `DataGeneratourSourceTest.testSequeunceCheckpointRestsore` where we test the restore functionality (and besides that, also the basic number generation for a range from 0 to 100).

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java (hunk 10 above)
Comment (xuzhiwen1255) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java:49]: I don't think so, because the current state is also out of order, because we can't guarantee the same processing speed of all subtasks.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java [hunk 11]
```diff
@@ -189,8 +206,28 @@ public static SequenceGenerator<String> stringGenerator(long start, long end) {
...
             }
         };
     }
+
+    /**
+     * The internal state of the sequence generator, which is used to record the latest state of the
+     * sequence value sent by the current sequence generator. When recovering from the state, it is
+     * guaranteed to continue sending the sequence value from the latest state.
+     */
+    private static class InternalState implements Comparable<InternalState> {
```
[the remark this answers is not in the document]
Comment (xuzhiwen1255) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java:219]: +1

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java (hunk 10 above)
Comment (xuzhiwen1255) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java:49]: And we use the priority queue to store state, so each subtask can be guaranteed to be sent according to the size of the sequence.

**Comment by xuzhiwen1255** [comment]
> I still think that it might make sense to introduce a `SequenceGeneratorTest`. This would allow us to test the edge cases easily (e.g. reaching `Long.MAX_VALUE`) more specifically.

In #21971 we added SequenceGeneratorTest, I think we can add the corresponding test method in #21971

@XComp WDYT?

**Comment by XComp** [comment]
> In https://github.com/apache/flink/pull/21971 we added SequenceGeneratorTest, I think we can add the corresponding test method in https://github.com/apache/flink/pull/21971

I'm not a big fan of adding tests that belong in a PR into a separate PR. You could just create the test class in this PR with the set of tests that fit to this PR. PR #21971 should then add additional tests related to its changes. The conflict resolution shouldn't be too complex because your adding independent test methods. :thinking:

**Comment by xuzhiwen1255** [comment]
> I'm not a big fan of adding tests that belong in a PR into a separate PR. You could just create the test class in this PR with the set of tests that fit to this PR. PR #21971 should then add additional tests related to its changes. The conflict resolution shouldn't be too complex because your adding independent test methods.

OK let me add a test.

**Comment by xuzhiwen1255** [comment]
> Any updates on that one @xuzhiwen1255 ? 🙂

@XComp  Sorry, the company has been busy iterating recently. I will update it in the next few days.

**Comment by XComp** [comment]
No worries. Especially considering that I won't be that responsive in September in general. I just wanted to let you know to make prioritization of work easier for you.

**Comment by xuzhiwen1255** [comment]
@XComp I'm sorry that I haven't had time to deal with this problem recently due to some things. Now I plan to continue to deal with this issue. Do you have time to take a look?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java [hunk 12]
```diff
@@ -66,33 +91,47 @@ public void open(
...
+                new ListStateDescriptor<>(
+                        name + "-sequence-state", TypeInformation.of(SubTaskState.class));
+        this.checkpointedState = context.getOperatorStateStore().getListState(stateDescriptor);
 
-            for (Long v : this.checkpointedState.get()) {
-                this.valuesToEmit.add(v);
-            }
+        if (!subTaskStates.isEmpty()) {
+            this.subTaskStates.clear();
+        }
```
Comment (XComp) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java:101]:
```suggestion
        this.subTaskStates.clear();
```
nit: clearing an empty collection shouldn't hurt.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java [hunk 13]
```diff
@@ -101,26 +140,20 @@ public void snapshotState(FunctionSnapshotContext context) throws Exception {
                 this.checkpointedState != null,
                 "The " + getClass().getSimpleName() + " state has not been properly initialized.");
 
-        this.checkpointedState.update(new ArrayList<>(this.valuesToEmit));
+        this.checkpointedState.clear();
+        this.checkpointedState.addAll(new ArrayList<>(subTaskStates));
```
Comment (XComp) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java:144]:
```suggestion
        this.checkpointedState.update(new ArrayList<>(subTaskStates));
```
nit: we can continue to rely on `update` here, I guess, to keep the diff smaller.

**Review by XComp — COMMENTED** [review]
Sorry for the delay. I found some time to look into the PR again. I have a few minor comments. The overall change looks good. But there are [CI test failures](https://dev.azure.com/apache-flink/apache-flink/_build/results?buildId=55868&view=logs&j=0c940707-2659-5648-cbe6-a1ad63045f0a&t=075c2716-8010-5565-fe08-3c4bb45824a4&l=8675) that seem to be related to the change. Can you have a look at it?

**Comment by xuzhiwen1255** [comment]
I think so. I will modify it as soon as possible.

**Comment by xuzhiwen1255** [comment]
@flinkbot run azure

**Comment by xuzhiwen1255** [comment]
Hi @XComp  I have solved the problem, please check again, thank you.
